drivers/i2c/i2c_rpi_pico_pio.c

```diff
@@ -0,0 +1,567 @@
+	.length = 19,
+	.origin = -1,
+};
+
+static inline pio_sm_config i2c_program_get_default_config(uint offset)
+{
+	pio_sm_config c = pio_get_default_sm_config();
+
+	sm_config_set_wrap(&c, offset + i2c_wrap_target, offset + i2c_wrap);
+	sm_config_set_sideset(&c, 2, true, true);
+	return c;
+}
+
+/* Instructions to manipulate clock and data lines to implement start/stop bits. */
+
+/*  0: set    pindirs, 0      side 0 [7] */
+#define I2C_SC0_SD0    0xf780
+/*  1: set    pindirs, 1      side 0 [7] */
+#define I2C_SC0_SD1    0xf781
+/*  2: set    pindirs, 0      side 1 [7] */
+#define I2C_SC1_SD0    0xff80
+/*  3: set    pindirs, 1      side 1 [7] */
+#define I2C_SC1_SD1    0xff81
+/*  4: wait   1 pin, 1 */
+#define I2C_WAIT_CLOCK 0x20a1
+
+/**@}*/
+
+/**@{*/
+
+#define PIO_I2C_ICOUNT_LSB 10
+#define PIO_I2C_FINAL_LSB  9
+#define PIO_I2C_DATA_LSB   1
+#define PIO_I2C_NAK_LSB    0
+
+static const uint16_t START_INSTS[] = {
+	1u << PIO_I2C_ICOUNT_LSB, /* Escape code for 2 instruction sequence */
+	I2C_SC1_SD0,              /* We are already in idle state, just pull SDA low */
+	I2C_SC0_SD0,              /* Also pull clock low so we can present data */
+};
+
+static const uint16_t STOP_INSTS[] = {
+	3u << PIO_I2C_ICOUNT_LSB, /* Escape code for 4 instruction sequence */
+	I2C_SC0_SD0,              /* SDA is unknown; pull it down */
+	I2C_SC1_SD0,              /* Release clock */
+	I2C_WAIT_CLOCK,           /* Wait for clock stretching */
+	I2C_SC1_SD1,              /* Release SDA to return to idle state */
+};
+
+static const uint16_t REPSTART_INSTS[] = {
+	4u << PIO_I2C_ICOUNT_LSB, /* Escape code for 5 instruction sequence */
+	I2C_SC0_SD1,              /* SDA is unknown; pull it down */
+	I2C_SC1_SD1,              /* Release clock */
+	I2C_WAIT_CLOCK,           /* Release wait for clock stretching */
+	I2C_SC1_SD0,              /* Pull SDA low */
+	I2C_SC0_SD0,              /* Pull clock low */
+};
+
+static bool pio_i2c_check_error(PIO pio, uint sm)
+{
+	return pio_interrupt_get(pio, sm);
+}
+
+static void pio_i2c_resume_after_error(PIO pio, uint sm)
+{
+	pio_sm_set_enabled(pio, sm, false);
+	pio_sm_drain_tx_fifo(pio, sm);
+	pio_sm_restart(pio, sm);
+	pio_sm_exec(pio, sm,
+		    (pio->sm[sm].execctrl & PIO_SM0_EXECCTRL_WRAP_BOTTOM_BITS) >>
+			    PIO_SM0_EXECCTRL_WRAP_BOTTOM_LSB);
+	pio_interrupt_clear(pio, sm);
+	pio_sm_set_enabled(pio, sm, true);
+}
+
+/**
+ * Enable autopush to RX fifo. The program is always reading, but if autopush is
+ * disabled the reads are not pushed out.
+ */
+static void pio_i2c_rx_enable(PIO pio, uint sm, bool en)
+{
+	if (en) {
+		hw_set_bits(&pio->sm[sm].shiftctrl, PIO_SM0_SHIFTCTRL_AUTOPUSH_BITS);
+	} else {
+		hw_clear_bits(&pio->sm[sm].shiftctrl, PIO_SM0_SHIFTCTRL_AUTOPUSH_BITS);
+	}
+}
+
+static inline void pio_i2c_put16(PIO pio, uint sm, uint16_t data)
+{
+	while (pio_sm_is_tx_fifo_full(pio, sm)) {
+		k_yield();
+	}
+	/* some versions of GCC dislike this */
+#ifdef __GNUC__
```

Comment:
Is this really necessary?

Reply:
Some of this is copy pasta'd from the i2c pio example code supplied by rpi. I'm happy to delete this until someone raises a need for it.